File: golang/core/mesh/mesh.go
// Package mesh implements the distributed inference mesh.
package mesh

import (
	"errors"
	"sort"
	"sync"
	"time"
)

// MeshConfig mirrors MeshConfig.
type MeshConfig struct {
	NodeID          string
	HeartbeatPeriod time.Duration
	ElectionTimeout time.Duration
	DiscoveryURL    string
	BindAddr        string
}

// DefaultMeshConfig returns sensible defaults.
func DefaultMeshConfig() MeshConfig {
	return MeshConfig{HeartbeatPeriod: 250 * time.Millisecond, ElectionTimeout: 1 * time.Second}
}

// MeshNode mirrors MeshNode.
type MeshNode struct {
	ID        string
	Addr      string
	Role      string
	LastSeen  time.Time
	Healthy   bool
	LayerHint int
}

// GossipRouter mirrors GossipRouter.
type GossipRouter struct {
	mu        sync.RWMutex
	peers     map[string]MeshNode
	callbacks []func(MeshNode)
}

// NewGossipRouter constructs an empty router.
func NewGossipRouter() *GossipRouter { return &GossipRouter{peers: map[string]MeshNode{}} }

// On registers a callback for peer changes.
func (r *GossipRouter) On(cb func(MeshNode)) { r.mu.Lock(); r.callbacks = append(r.callbacks, cb); r.mu.Unlock() }

// Update inserts or refreshes a peer.
func (r *GossipRouter) Update(n MeshNode) {
	r.mu.Lock()
	n.LastSeen = time.Now()
	r.peers[n.ID] = n
	cbs := r.callbacks
	r.mu.Unlock()
	for _, cb := range cbs {
		cb(n)
	}
}

// Peers returns a copy of the current peer list.
func (r *GossipRouter) Peers() []MeshNode {
	r.mu.RLock()
	defer r.mu.RUnlock()
	out := make([]MeshNode, 0, len(r.peers))
	for _, p := range r.peers {
		out = append(out, p)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].ID < out[j].ID })
	return out
}

// PeerByID returns the named peer.
func (r *GossipRouter) PeerByID(id string) (MeshNode, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	n, ok := r.peers[id]
	return n, ok
}

// Remove removes a peer.
func (r *GossipRouter) Remove(id string) {
	r.mu.Lock()
	delete(r.peers, id)
	r.mu.Unlock()
}

// RingTransport mirrors RingTransport.
type RingTransport struct {
	mu    sync.Mutex
	peers []string
}

// NewRingTransport constructs a ring with the given peer addresses.
func NewRingTransport(peers []string) *RingTransport { return &RingTransport{peers: peers} }

// Next returns the next peer in the ring.
func (r *RingTransport) Next(self string) string {
	r.mu.Lock()
	defer r.mu.Unlock()
	if len(r.peers) == 0 {
		return ""
	}
	for i, p := range r.peers {
		if p == self {
			return r.peers[(i+1)%len(r.peers)]
		}
	}
	return r.peers[0]
}

// ChannelTransport mirrors ChannelTransport. It uses Go channels for tests.
type ChannelTransport struct {
	In  chan []byte
	Out chan []byte
}

// NewChannelTransport constructs an in-memory transport.
func NewChannelTransport() *ChannelTransport { return &ChannelTransport{In: make(chan []byte, 64), Out: make(chan []byte, 64)} }

// Send writes a message.
func (c *ChannelTransport) Send(msg []byte) { c.Out <- msg }

// Recv returns a message or nil.
func (c *ChannelTransport) Recv() []byte {
	select {
	case m := <-c.In:
		return m
	default:
		return nil
	}
}

// TcpTransport mirrors TcpTransport. It is a thin shell that records
// configuration but does not actually open TCP connections.
type TcpTransport struct {
	Addr string
}

// NewTcpTransport constructs a transport that will bind to `addr`.
func NewTcpTransport(addr string) *TcpTransport { return &TcpTransport{Addr: addr} }

// ShardPlan mirrors ShardPlan.
type ShardPlan struct {
	Shards       []MeshShard
	LayerRange   [2]int
	ReplicaIndex int
}

// MeshShard mirrors MeshShard.
type MeshShard struct {
	NodeID     string
	LayerStart int
	LayerEnd   int
	DeviceID   int
	Backend    string
}

// Validate ensures the shard plan is consistent.
func (p *ShardPlan) Validate() error {
	if len(p.Shards) == 0 {
		return errors.New("mesh: empty shard plan")
	}
	if p.LayerRange[0] >= p.LayerRange[1] {
		return errors.New("mesh: invalid layer range")
	}
	for _, s := range p.Shards {
		if s.LayerStart >= s.LayerEnd {
			return errors.New("mesh: invalid shard range")
		}
	}
	return nil
}

// DiscoveryService mirrors DiscoveryService.
type DiscoveryService struct {
	mu      sync.Mutex
	router  *GossipRouter
	known   map[string]struct{}
}

// NewDiscoveryService constructs a service bound to a router.
func NewDiscoveryService(r *GossipRouter) *DiscoveryService { return &DiscoveryService{router: r, known: map[string]struct{}{}} }

// Announce registers a node with the service.
func (d *DiscoveryService) Announce(n MeshNode) {
	d.mu.Lock()
	d.known[n.ID] = struct{}{}
	d.mu.Unlock()
	d.router.Update(n)
}

// Forget removes a node.
func (d *DiscoveryService) Forget(id string) {
	d.mu.Lock()
	delete(d.known, id)
	d.mu.Unlock()
	d.router.Remove(id)
}

// BullyElection mirrors BullyElection.
type BullyElection struct {
	mu     sync.Mutex
	peers  []string
	leader string
}

// NewBullyElection constructs a bully election for the given peer ids.
func NewBullyElection(peers []string) *BullyElection {
	return &BullyElection{peers: peers, leader: peers[0]}
}

// Leader returns the current leader id.
func (b *BullyElection) Leader() string {
	b.mu.Lock()
	defer b.mu.Unlock()
	return b.leader
}

// SetLeader manually assigns a leader.
func (b *BullyElection) SetLeader(id string) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.leader = id
}

// TopologyGraph mirrors TopologyGraph.
type TopologyGraph struct {
	mu    sync.Mutex
	nodes map[string]MeshNode
	edges map[string][]string
}

// NewTopologyGraph constructs an empty graph.
func NewTopologyGraph() *TopologyGraph { return &TopologyGraph{nodes: map[string]MeshNode{}, edges: map[string][]string{}} }

// AddNode inserts a node.
func (g *TopologyGraph) AddNode(n MeshNode) {
	g.mu.Lock()
	g.nodes[n.ID] = n
	if _, ok := g.edges[n.ID]; !ok {
		g.edges[n.ID] = []string{}
	}
	g.mu.Unlock()
}

// AddEdge records a link from a -> b.
func (g *TopologyGraph) AddEdge(a, b string) {
	g.mu.Lock()
	g.edges[a] = append(g.edges[a], b)
	g.mu.Unlock()
}

// Neighbors returns the neighbors of a node.
func (g *TopologyGraph) Neighbors(id string) []string {
	g.mu.Lock()
	defer g.mu.Unlock()
	out := append([]string(nil), g.edges[id]...)
	sort.Strings(out)
	return out
}

// Nodes returns all nodes.
func (g *TopologyGraph) Nodes() []MeshNode {
	g.mu.Lock()
	defer g.mu.Unlock()
	out := make([]MeshNode, 0, len(g.nodes))
	for _, n := range g.nodes {
		out = append(out, n)
	}
	return out
}

// MeshChatEngine mirrors MeshChatEngine.
type MeshChatEngine struct {
	mu        sync.Mutex
	Router    *GossipRouter
	Shards    map[string]*ShardPlan
	LocalNode MeshNode
}

// NewMeshChatEngine constructs a new engine.
func NewMeshChatEngine(local MeshNode) *MeshChatEngine {
	return &MeshChatEngine{Router: NewGossipRouter(), Shards: map[string]*ShardPlan{}, LocalNode: local}
}

// RegisterShard registers a shard plan under a key.
func (e *MeshChatEngine) RegisterShard(key string, plan *ShardPlan) {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.Shards[key] = plan
}

// LookupShard returns the named shard plan.
func (e *MeshChatEngine) LookupShard(key string) (*ShardPlan, bool) {
	e.mu.Lock()
	defer e.mu.Unlock()
	s, ok := e.Shards[key]
	return s, ok
}

// LoadProgressReport mirrors LoadProgressReport.
type LoadProgressReport struct {
	NodeID        string
	BytesLoaded   int64
	BytesTotal    int64
	PercentDone   float32
	LayersReady   int
	LayersTotal   int
	ElapsedMillis int64
}

// MeshValidationReport mirrors MeshValidationReport.
type MeshValidationReport struct {
	TotalNodes   int
	HealthyNodes int
	LayersCovered int
	MissingShards []string
}

// Validate checks a shard plan against a list of healthy nodes.
func Validate(plan *ShardPlan, healthy []MeshNode) MeshValidationReport {
	rep := MeshValidationReport{TotalNodes: len(healthy), HealthyNodes: 0}
	healthyIDs := map[string]struct{}{}
	for _, n := range healthy {
		if n.Healthy {
			rep.HealthyNodes++
		}
		healthyIDs[n.ID] = struct{}{}
	}
	if plan == nil {
		return rep
	}
	for _, s := range plan.Shards {
		if _, ok := healthyIDs[s.NodeID]; ok {
			rep.LayersCovered += s.LayerEnd - s.LayerStart
		} else {
			rep.MissingShards = append(rep.MissingShards, s.NodeID)
		}
	}
	return rep
}

File: oxidize-python/oxidize_python/core/mesh/mesh.py
"""Distributed mesh mirroring oxidize-golang/core/mesh/mesh.go."""

from __future__ import annotations

import threading
import time
from dataclasses import dataclass, field


@dataclass
class MeshConfig:
    node_id: str = ""
    heartbeat_period: float = 0.25
    election_timeout: float = 1.0
    discovery_url: str = ""
    bind_addr: str = ""


def default_mesh_config() -> MeshConfig:
    return MeshConfig()


@dataclass
class MeshNode:
    id: str
    addr: str = ""
    role: str = ""
    last_seen: float = 0.0
    healthy: bool = True
    layer_hint: int = 0


class GossipRouter:
    def __init__(self) -> None:
        self._mu = threading.RLock()
        self._peers: dict[str, MeshNode] = {}
        self._callbacks: list = []

    def on(self, cb) -> None:
        with self._mu:
            self._callbacks.append(cb)

    def update(self, node: MeshNode) -> None:
        with self._mu:
            node.last_seen = time.time()
            self._peers[node.id] = node
            cbs = list(self._callbacks)
        for cb in cbs:
            cb(node)

    def peers(self) -> list[MeshNode]:
        with self._mu:
            out = list(self._peers.values())
        return sorted(out, key=lambda n: n.id)

    def peer_by_id(self, node_id: str) -> tuple[MeshNode, bool]:
        with self._mu:
            n = self._peers.get(node_id)
        return (n, n is not None) if n else (MeshNode(""), False)

    def remove(self, node_id: str) -> None:
        with self._mu:
            self._peers.pop(node_id, None)


class RingTransport:
    def __init__(self, peers: list[str]) -> None:
        self._mu = threading.Lock()
        self._peers = list(peers)

    def next(self, self_addr: str) -> str:
        with self._mu:
            if not self._peers:
                return ""
            for i, p in enumerate(self._peers):
                if p == self_addr:
                    return self._peers[(i + 1) % len(self._peers)]
            return self._peers[0]


class ChannelTransport:
    def __init__(self) -> None:
        import queue
        self.in_q: queue.Queue[bytes] = queue.Queue(maxsize=64)
        self.out_q: queue.Queue[bytes] = queue.Queue(maxsize=64)

    def send(self, msg: bytes) -> None:
        self.out_q.put(msg)

    def recv(self) -> bytes | None:
        try:
            return self.in_q.get_nowait()
        except Exception:
            try:
                return self.out_q.get_nowait()
            except Exception:
                return None


@dataclass
class TcpTransport:
    addr: str


@dataclass
class MeshShard:
    node_id: str
    layer_start: int
    layer_end: int
    device_id: int = 0
    backend: str = ""


@dataclass
class ShardPlan:
    shards: list[MeshShard] = field(default_factory=list)
    layer_range: tuple[int, int] = (0, 0)
    replica_index: int = 0

    def validate(self) -> None:
        if not self.shards:
            raise ValueError("mesh: empty shard plan")
        if self.layer_range[0] >= self.layer_range[1]:
            raise ValueError("mesh: invalid layer range")
        for s in self.shards:
            if s.layer_start >= s.layer_end:
                raise ValueError("mesh: invalid shard range")


class DiscoveryService:
    def __init__(self, router: GossipRouter) -> None:
        self._mu = threading.Lock()
        self.router = router
        self._known: set[str] = set()

    def announce(self, node: MeshNode) -> None:
        with self._mu:
            self._known.add(node.id)
        self.router.update(node)

    def forget(self, node_id: str) -> None:
        with self._mu:
            self._known.discard(node_id)
        self.router.remove(node_id)


class BullyElection:
    def __init__(self, peers: list[str]) -> None:
        self._mu = threading.Lock()
        self._peers = peers
        self._leader = peers[0] if peers else ""

    def leader(self) -> str:
        with self._mu:
            return self._leader

    def set_leader(self, leader_id: str) -> None:
        with self._mu:
            self._leader = leader_id


class TopologyGraph:
    def __init__(self) -> None:
        self._mu = threading.Lock()
        self._nodes: dict[str, MeshNode] = {}
        self._edges: dict[str, list[str]] = {}

    def add_node(self, node: MeshNode) -> None:
        with self._mu:
            self._nodes[node.id] = node
            self._edges.setdefault(node.id, [])

    def add_edge(self, a: str, b: str) -> None:
        with self._mu:
            self._edges.setdefault(a, []).append(b)

    def neighbors(self, node_id: str) -> list[str]:
        with self._mu:
            return sorted(self._edges.get(node_id, []))

    def nodes(self) -> list[MeshNode]:
        with self._mu:
            return list(self._nodes.values())


@dataclass
class LoadProgressReport:
    node_id: str = ""
    bytes_loaded: int = 0
    bytes_total: int = 0
    percent_done: float = 0.0
    layers_ready: int = 0
    layers_total: int = 0
    elapsed_millis: int = 0


@dataclass
class MeshValidationReport:
    total_nodes: int = 0
    healthy_nodes: int = 0
    layers_covered: int = 0
    missing_shards: list[str] = field(default_factory=list)


class MeshChatEngine:
    def __init__(self, local: MeshNode) -> None:
        self._mu = threading.Lock()
        self.router = GossipRouter()
        self.shards: dict[str, ShardPlan] = {}
        self.local_node = local

    def register_shard(self, key: str, plan: ShardPlan) -> None:
        with self._mu:
            self.shards[key] = plan

    def lookup_shard(self, key: str) -> tuple[ShardPlan | None, bool]:
        with self._mu:
            p = self.shards.get(key)
        return (p, p is not None)


def validate(plan: ShardPlan | None, healthy: list[MeshNode]) -> MeshValidationReport:
    rep = MeshValidationReport(total_nodes=len(healthy))
    healthy_ids = set()
    for n in healthy:
        if n.healthy:
            rep.healthy_nodes += 1
        healthy_ids.add(n.id)
    if plan is None:
        return rep
    for s in plan.shards:
        if s.node_id in healthy_ids:
            rep.layers_covered += s.layer_end - s.layer_start
        else:
            rep.missing_shards.append(s.node_id)
    return rep
